sdk/client/src/stream/websocket_transport.rs

```diff
@@ -0,0 +1,156 @@
+            .body(())
+            .map_err(StreamError::from_http_error)?;
+
+        let (stream, _) = connect_async_with_config(request, websocket_config)
+            .await
+            .map_err(StreamError::from_tungstenite_error)?;
+
+        let (sink, stream) = stream.split();
+
+        Ok(Self {
+            stream: Some(stream),
+            sink,
+            next_id: AtomicU64::new(0),
+            channel_task: None,
+        })
+    }
+
+    pub async fn send(&mut self, request_json: String) -> StreamResult<()> {
+        self.sink
+            .send(Message::text(request_json))
+            .await
+            .map_err(StreamError::encode)?;
+        Ok(())
+    }
+
+    pub async fn send_method_request(
+        &mut self,
+        request: StreamMethodRequest,
+        id: Option<Id>,
+    ) -> StreamResult<Id> {
+        let id = id.unwrap_or_else(|| self.get_next_id());
+        let request = StreamJsonRpcRequest::new(request, id.clone());
+        self.send_request(&request).await
+    }
+
+    pub async fn send_request(&mut self, request: &StreamJsonRpcRequest) -> StreamResult<Id> {
+        let json = serde_json::to_string(&request)?;
+        self.send(json).await?;
+        Ok(request.id.clone())
+    }
+
+    pub fn get_next_id(&self) -> Id {
+        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+        Id::Number(id)
+    }
+
+    /// This consumes the websocket stream (but not the sink)
+    /// Possibilities:
+    /// 1. `stream.next().await` returns `None`: this happens if a stream is closed
+    /// 2. `stream.next().await` returns `Some(Err)`: this is generally a transport error, and the
+    ///     stream can be considered closed
+    /// 3. `stream.next().await` returns `Some(Ok(Message))`: this may or may not be valid JSON from
+    ///     the server, so we need to do a bit of validation:
+    ///     1. `msg.is_text()` must be `true`: we ignore `Ping`/`Pong`, `Binary`, etc message types
+    ///     2. `msg.to_text()` must be `Ok(String)`: Ensures the message is valid UTF8
+    ///     3. `StreamJsonRpcResponse::from_str(msg)` must be `Ok(StreamJsonRpcResponse)`: this is
+    ///         serde deserialization, which validates that the JSON is well formed, and matches the
+    ///         `StreamJsonRpcResponse` struct
+    ///
+    pub fn get_stream(mut self) -> (StreamingClientReceiver, Self) {
+        let (sender, receiver) = mpsc::channel(100);
+
+        let mut stream = self
+            .stream
+            .expect("Stream is `None`: it has already been consumed");
+        self.stream = None;
+
+        self.channel_task = Some(tokio::task::spawn(async move {
+            loop {
+                match stream.next().await {
+                    None => {
+                        sender
+                            .send(Err(StreamError::connection_closed(None::<StreamError>)))
+                            .await
+                            .ok();
+                        continue;
```

Comment:
I don't think this does anything?

Reply:
when the upstream websocket connection is closed, it sends a message to the client so that the client can handle the shutdown in a nice way on their end